operator/v1/types_ingress.go

```diff
@@ -235,8 +235,50 @@ type IngressControllerSpec struct {
 	// +nullable
 	// +kubebuilder:pruning:PreserveUnknownFields
 	UnsupportedConfigOverrides runtime.RawExtension `json:"unsupportedConfigOverrides"`
+
+	// httpCompression defines a policy for HTTP traffic compression.
+	// By default, there is no HTTP compression.
+	//
+	// +optional
+	HTTPCompression HTTPCompressionPolicy `json:"httpCompression,omitempty"`
 }
 
+// httpCompressionPolicy turns on compression for the specified MIME types.
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the specified
+// MIME types.
+type HTTPCompressionPolicy struct {
+	// compressionMIMETypes is a list of MIME types that should have compression applied.
+	// This list can be empty, in which case the ingress controller does not apply compression.
+	//
+	// Note: Not all MIME types benefit from compression, but HAProxy will still use resources
+	// to try to compress if instructed to.  Generally speaking, text (html, css, js, etc.)
+	// formats benefit from compression, but formats that are already compressed (image,
+	// audio, video, etc.) benefit little in exchange for the time and cpu spent on compressing
+	// again. See https://joehonton.medium.com/the-gzip-penalty-d31bd697f1a2
+	//
+	// +required
+	// +kubebuilder:validation:MinItems=0
+	CompressionMIMETypes []CompressionMIMEType `json:"mimeTypes,omitempty"`
+}
+
+// CompressionMIMEType defines the format of a single MIME type.
+// E.g. "text/css;charset=utf-8", "text/html", "text/*", "image/svg+xml",
+// "application/octet-stream", "X-custom/customsub", etc.
+// The format should loosely follow the Content-Type definition in RFC1341.
+// For the type in the <type>/<subtype> MIME type format, it can be one of:
+// application, audio, image, message, multipart, text, video, or a custom
+// type preceded by X- and not using a space or any of the characters
+// in the set ()[]<>@,;:\"/?.=
+// For the subtype in the <type>/<subtype> MIME type format, there are no
+// restrictions.
+//
+// +kubebuilder:validation:Pattern=`^"?(X-[^ ()[\]<>@,;:"/?.=]+|(application|audio|image|message|multipart|text|video))"?/"?(.+)+"?$`
```

Comment:
* The cited RFC says custom types should use capital "X-", but some other RFCs use lower-case "x-", and I think most examples I've seen in the wild are lower-case as well; we probably should add `(?i)` to make the pattern case-insensitive.    
* A character class that includes brackets should put the close bracket first to avoid ambiguity.  
* Generally, you should avoid using `(.+)+` in regexps because it can cause some regexp implementations to explode when backtracking.  Granted, it might not cause harm in this case because `.+` is greedy and the rest of the pattern can match the empty string, so it should never backtrack (by the same token, the `+` and the `"?` at the end of the pattern are superfluous).  Also, I forget how Go's RE2 implementation handles that sort of thing, and maybe this isn't an issue there.  However, it's good form to avoid `(.+)+` and the possible resultant combinatorial explosion.
* The RFC says that  parameter values may be quoted, but otherwise double-quotes are not allowed.  
* You could make the `.+` a little more specific to allow only valid subtypes and parameters.  (The suggested pattern is a little messy; for purposes of reading it, replacing `[^][ ()\\<>@,;:"/?.=]+` with `<token>` will help.)
```suggestion
// +kubebuilder:validation:Pattern=`^(?i)(x-[^][ ()\\<>@,;:"/?.=]+|application|audio|image|message|multipart|text|video)/[^][ ()\\<>@,;:"/?.=]+(;[^][ ()\\<>@,;:"/?.=]+=([^][ ()\\<>@,;:"/?.=]+|"(\\.|[^\x0D"\\])*"))*$`
```

Reply:
It still needs tweaking because this doesn't seem to allow the following valid patterns:
text/*
text/css; charset=utf-8
x-application/"pdf; a=b"

Reply:
This pattern has been reworked based on your comments.  The pattern can be exercised here: https://play.golang.org/p/Wz8TAPmlnH9